File: src/theme.c
#include "theme.h"
#include <commctrl.h>
#include <uxtheme.h>
#include <dwmapi.h>

#ifndef DWMWA_USE_IMMERSIVE_DARK_MODE
#define DWMWA_USE_IMMERSIVE_DARK_MODE 20
#endif

#define DARK_BG   RGB(32, 32, 32)
#define DARK_TEXT RGB(255, 255, 255)

static BOOL   g_dark = FALSE;
static HBRUSH g_dark_brush = NULL;

void theme_update(void) {
	DWORD value = 1, size = sizeof(value);
	RegGetValueW(HKEY_CURRENT_USER, L"Software\\Microsoft\\Windows\\CurrentVersion\\Themes\\Personalize", L"AppsUseLightTheme", RRF_RT_REG_DWORD, NULL, &value, &size);
	g_dark = (value == 0);
}

BOOL theme_is_dark(void) {
	return g_dark;
}

void theme_apply_titlebar(HWND hwnd) {
	BOOL dark = g_dark;
	DwmSetWindowAttribute(hwnd, DWMWA_USE_IMMERSIVE_DARK_MODE, &dark, sizeof(BOOL));
}

void theme_apply_listview(HWND hwnd_list) {
	if (g_dark) {
		SetWindowTheme(hwnd_list, L"DarkMode_Explorer", NULL);
		ListView_SetBkColor(hwnd_list, DARK_BG);
		ListView_SetTextBkColor(hwnd_list, DARK_BG);
		ListView_SetTextColor(hwnd_list, DARK_TEXT);
	} else {
		SetWindowTheme(hwnd_list, L"Explorer", NULL);
		ListView_SetBkColor(hwnd_list, CLR_DEFAULT);
		ListView_SetTextBkColor(hwnd_list, CLR_DEFAULT);
		ListView_SetTextColor(hwnd_list, CLR_DEFAULT);
	}
	InvalidateRect(hwnd_list, NULL, TRUE);
}

void theme_apply_button(HWND hwnd) {
	SetWindowTheme(hwnd, g_dark ? L"DarkMode_Explorer" : L"Explorer", NULL);
}

HBRUSH theme_ctl_color(HDC hdc) {
	if (!g_dark) return NULL;
	if (!g_dark_brush) g_dark_brush = CreateSolidBrush(DARK_BG);
	SetTextColor(hdc, DARK_TEXT);
	SetBkColor(hdc, DARK_BG);
	return g_dark_brush;
}

HBRUSH theme_bg_brush(void) {
	if (!g_dark) return NULL;
	if (!g_dark_brush) g_dark_brush = CreateSolidBrush(DARK_BG);
	return g_dark_brush;
}

File: src/theme.zig
const std = @import("std");
const win32 = @import("win32.zig");
const L = std.unicode.utf8ToUtf16LeStringLiteral;

fn rgb(r: u8, g: u8, b: u8) win32.COLORREF {
	return @as(u32, r) | (@as(u32, g) << 8) | (@as(u32, b) << 16);
}

const DARK_BG: win32.COLORREF = rgb(32, 32, 32);
const DARK_TEXT: win32.COLORREF = rgb(255, 255, 255);
const CLR_DEFAULT: win32.COLORREF = 0xFF000000;
const COLOR_NONE: win32.COLORREF = 0xFFFFFFFF;
const DWMWA_USE_IMMERSIVE_DARK_MODE: win32.DWORD = 20;

fn colorParam(c: win32.COLORREF) win32.LPARAM {
	return @intCast(c);
}

const LVM_SETBKCOLOR: win32.UINT = 0x1000 + 1;
const LVM_SETTEXTCOLOR: win32.UINT = 0x1000 + 36;
const LVM_SETTEXTBKCOLOR: win32.UINT = 0x1000 + 38;
const TVM_SETBKCOLOR: win32.UINT = 0x1100 + 29;
const TVM_SETTEXTCOLOR: win32.UINT = 0x1100 + 30;

var g_dark: win32.BOOL = 0;
var g_dark_brush: win32.HBRUSH = null;

pub fn update() void {
	var value: win32.DWORD = 1;
	var size: win32.DWORD = @sizeOf(win32.DWORD);
	_ = win32.RegGetValueW(win32.HKEY_CURRENT_USER, L("Software\\Microsoft\\Windows\\CurrentVersion\\Themes\\Personalize"), L("AppsUseLightTheme"), win32.RRF_RT_REG_DWORD, null, &value, &size);
	g_dark = if (value == 0) 1 else 0;
}

pub fn isDark() win32.BOOL {
	return g_dark;
}

pub fn applyTitlebar(hwnd: win32.HWND) void {
	const dark = g_dark;
	_ = win32.DwmSetWindowAttribute(hwnd, DWMWA_USE_IMMERSIVE_DARK_MODE, &dark, @sizeOf(win32.BOOL));
}

pub fn applyListview(hwnd_list: win32.HWND) void {
	if (g_dark != 0) {
		_ = win32.SetWindowTheme(hwnd_list, L("DarkMode_Explorer"), null);
		_ = win32.SendMessageW(hwnd_list, LVM_SETBKCOLOR, 0, colorParam(DARK_BG));
		_ = win32.SendMessageW(hwnd_list, LVM_SETTEXTBKCOLOR, 0, colorParam(DARK_BG));
		_ = win32.SendMessageW(hwnd_list, LVM_SETTEXTCOLOR, 0, colorParam(DARK_TEXT));
	} else {
		_ = win32.SetWindowTheme(hwnd_list, L("Explorer"), null);
		_ = win32.SendMessageW(hwnd_list, LVM_SETBKCOLOR, 0, colorParam(CLR_DEFAULT));
		_ = win32.SendMessageW(hwnd_list, LVM_SETTEXTBKCOLOR, 0, colorParam(CLR_DEFAULT));
		_ = win32.SendMessageW(hwnd_list, LVM_SETTEXTCOLOR, 0, colorParam(CLR_DEFAULT));
	}
	_ = win32.InvalidateRect(hwnd_list, null, 1);
}

pub fn applyButton(hwnd: win32.HWND) void {
	_ = win32.SetWindowTheme(hwnd, if (g_dark != 0) L("DarkMode_Explorer") else L("Explorer"), null);
}

pub fn applyTreeview(hwnd_tree: win32.HWND) void {
	if (g_dark != 0) {
		_ = win32.SetWindowTheme(hwnd_tree, L("DarkMode_Explorer"), null);
		_ = win32.SendMessageW(hwnd_tree, TVM_SETBKCOLOR, 0, colorParam(DARK_BG));
		_ = win32.SendMessageW(hwnd_tree, TVM_SETTEXTCOLOR, 0, colorParam(DARK_TEXT));
	} else {
		_ = win32.SetWindowTheme(hwnd_tree, L("Explorer"), null);
		_ = win32.SendMessageW(hwnd_tree, TVM_SETBKCOLOR, 0, colorParam(COLOR_NONE));
		_ = win32.SendMessageW(hwnd_tree, TVM_SETTEXTCOLOR, 0, colorParam(COLOR_NONE));
	}
	_ = win32.InvalidateRect(hwnd_tree, null, 1);
}

pub fn ctlColor(hdc: win32.HDC) win32.HBRUSH {
	if (g_dark == 0) return null;
	if (g_dark_brush == null) g_dark_brush = win32.CreateSolidBrush(DARK_BG);
	_ = win32.SetTextColor(hdc, DARK_TEXT);
	_ = win32.SetBkColor(hdc, DARK_BG);
	return g_dark_brush;
}

pub fn bgBrush() win32.HBRUSH {
	if (g_dark == 0) return null;
	if (g_dark_brush == null) g_dark_brush = win32.CreateSolidBrush(DARK_BG);
	return g_dark_brush;
}
